### PR TITLE
Added to_lower, to_upper, reverse and to_title function to stdlib_string_type.f90 file

### LINKED ISSUE #335: Implement to_lower, to_upper, to_title and reverse for string_type
The `stdlib_string_type` currently only implements the bare minimum functionality. To start extending the functionality the functions `to_lower`, `to_upper`, `to_title` and `reverse` implemented in `stdlib_ascii` should be implemented for `stdlib_string_type` as well.

### PR DESCRIPTION
**Pull request by Aman-Godara**
**Status**: open for review.

**Implemented**: the functions asked in #335, added unit tests for each of these functions in `test_string_functions.f90` file and documented the functions in `stdlib_string_type.md` file. (Fixes #335)
**Yet to be done**: https://github.com/fortran-lang/stdlib/pull/346#issuecomment-802004187
**Note**: Please have a look at the recent [suggestion comment ](https://github.com/fortran-lang/stdlib/issues/335#issuecomment-798815164)concerning `to_title` function

I am still trying to get the issues concerning my compiler fixed to run the code on my machine. The code written is **not** checked to run successfully on machine yet. Thus, I kindly request the reviewers to please also run the code once.

### CODE REVIEW

File: src/stdlib_string_type.f90 [hunk 1]
```diff
@@ -12,11 +12,13 @@
 !>
 !> The specification of this module is available [here](../page/specs/stdlib_string_type.html).
 module stdlib_string_type
+    use stdlib_ascii, only: to_lower, to_upper, to_title, reverse
```
Comment (awvwgk) [inline src/stdlib_string_type.f90:15]: Avoid namespace pollution:

```suggestion
    use stdlib_ascii, only: to_lower_ => to_lower, to_upper_ => to_upper, to_title_ => to_title, reverse_ => reverse
```

File: src/stdlib_string_type.f90 [hunk 2]
```diff
@@ -447,6 +477,46 @@ elemental function repeat_string(string, ncopies) result(repeated_string)
     end function repeat_string
 
 
+    !> Convert the character sequence hold by the input string to lower case
+    elemental function to_lower_string(string) result(lowered_string)
+        type(string_type), intent(in) :: string
+        type(string_type) :: lowered_string
+
+        lowered_string%raw = to_lower(string%raw)
```
Comment (awvwgk) [inline src/stdlib_string_type.f90:485]:
```suggestion
        lowered_string%raw = to_lower_(maybe(string))
```

File: src/stdlib_string_type.f90 [hunk 3]
```diff
@@ -447,6 +477,46 @@ elemental function repeat_string(string, ncopies) result(repeated_string)
...
+
+    end function to_lower_string
+
+
+    !> Convert the character sequence hold by the input string to upper case
+    elemental function to_upper_string(string) result(uppered_string)
+        type(string_type), intent(in) :: string
+        type(string_type) :: uppered_string
+
+        uppered_string%raw = to_upper(string%raw)
```
Comment (awvwgk) [inline src/stdlib_string_type.f90:495]:
```suggestion
        uppered_string%raw = to_upper_(maybe(string))
```

File: src/stdlib_string_type.f90 [hunk 4]
```diff
@@ -447,6 +477,46 @@ elemental function repeat_string(string, ncopies) result(repeated_string)
...
+
+    end function to_upper_string
+
+
+    !> Convert the character sequence hold by the input string to title case
+    elemental function to_title_string(string) result(titled_string)
+        type(string_type), intent(in) :: string
+        type(string_type) :: titled_string
+
+        titled_string%raw = to_title(string%raw)
```
Comment (awvwgk) [inline src/stdlib_string_type.f90:505]:
```suggestion
        titled_string%raw = to_title_(maybe(string))
```

File: src/stdlib_string_type.f90 [hunk 5]
```diff
@@ -447,6 +477,46 @@ elemental function repeat_string(string, ncopies) result(repeated_string)
...
+
+    end function to_title_string
+
+
+    !> Reverse the character sequence hold by the input string
+    elemental function reverse_string(string) result(reversed_string)
+        type(string_type), intent(in) :: string
+        type(string_type) :: reversed_string
+
+        reversed_string%raw = reverse(string%raw)
```
Comment (awvwgk) [inline src/stdlib_string_type.f90:515]:
```suggestion
        reversed_string%raw = reverse_(maybe(string))
```

**Review by awvwgk — COMMENTED** [review]
Thanks for sharing. I suggest to rename the imports from `stdlib_ascii` to avoid polluting the namespace of `stdlib_string_type` and reexporting them accidentally.

I recommend to use the `maybe()` wrapper for safe memory access inside `stdlib_string_type` rather than directly accessing the raw value on the right-hand-side.

**Comment by awvwgk** [comment]
> Should the newly created functions in stdlib_string_type.f90 module (to_lower_string, to_upper_string, etc.) be elemental or pure?

Elemental should be the way to go here. For `stdlib_ascii` we discussed the usage of `pure` vs `elemental` briefly here: https://github.com/fortran-lang/stdlib/pull/310#issuecomment-776978642. Those concerns are mainly related to the intrinsic character type and not the `string_type`.

Also the manual Makefile failure is related to the newly introduced dependency of `stdlib_string_type` on `stdlib_ascii` and can be fixed by adding
```make
stdlib_string_type.o: stdlib_ascii.o
```
to `src/Makefile.manual`.

**Comment by Aman-Godara** [comment]
I will do the rest of the changes in the pull request as soon as my minor exams (15th to 18th) are over. Nevertheless, I remain active on github to check for the updates. I will start working on this issue full time from 18th until. If you anyone have any suggestion, please post them here. Thank you.
I will delete this comment once it becomes irrelevant.

**Comment by Aman-Godara** [comment]
Can I create a new .f90 file in the `src/tests/string` to create test cases for the functions implemented above or should I edit an already existing file in the folder.

**Comment by awvwgk** [comment]
> Can I create a new .f90 file in the `src/tests/string` to create test cases for the functions implemented above or should I edit an already existing file in the folder.

You can add a new file, but make sure to register the unit test in the CMake build file in the same directory than as well.

**Comment by Aman-Godara** [comment]
What happens when a user gives a `string_type` object on the right hand side instead of a character sequence while handling the `assignment operator(=)` with `string_type` variables.

**Review by milancurcic — APPROVED** [review]
Looks good to me, thanks @Aman-Godara!

**Comment by milancurcic** [comment]
> What happens when a user gives a string_type object on the right hand side instead of a character sequence while handling the assignment operator(=) with string_type variables. I think current implementation doesn't handle such possibility.

Assigning `string_type` to another `string_type` comes out-of-the box with the derived type. It should just work. Is that what you mean or something else?

**Review by awvwgk — COMMENTED** [review]
@Aman-Godara Can you add the new procedures to the specification of the `string_type` in `doc/specs/stdlib_string_type.md` as well?

**Comment by Aman-Godara** [comment]
@awvwgk Yes, I am working on documenting the functions created. I just needed a little help with https://github.com/fortran-lang/stdlib/pull/346#issuecomment-802004187 to proceed further because I am at the starting stage of learning CMake.

@milancurcic yes, you got it right. I wanted to talk more about it, but before that let me try to figure things out on my own by learning fortran.

File: src/stdlib_string_type.f90 [hunk 6]
```diff
@@ -12,11 +12,15 @@
...
+                            to_title_ => to_title, reverse_ => reverse
+
     implicit none
     private
 
     public :: string_type
-    public :: len, len_trim, trim, index, scan, verify, repeat, adjustr, adjustl
+    public :: len, len_trim, trim, index, scan, verify, repeat, adjustr, &
+              adjustl, to_lower, to_upper, to_title, reverse
     public :: lgt, lge, llt, lle, char, ichar, iachar
```
Comment (awvwgk) [inline src/stdlib_string_type.f90:24]: Just a bit of reordering here to keep the overloaded intrinsics together.
```suggestion
    public :: len, len_trim, trim, index, scan, verify, repeat, adjustr, adjustl
    public :: lgt, lge, llt, lle, char, ichar, iachar
    public :: to_lower, to_upper, to_title, reverse
```

**Review by awvwgk — APPROVED** [review]
Thanks, this looks good now.

**Comment by milancurcic** [comment]
With two approvals, I will merge. Thank you, @Aman-Godara!

**Comment by Aman-Godara** [comment]
> I recommend to use the `maybe()` wrapper for safe memory access inside `stdlib_string_type` rather than directly accessing the raw value on the right-hand-side.

@awvwgk does `maybe()` wrapper handle the case when `raw` (component of `string_type` derived type) is not allocated?

**Comment by awvwgk** [comment]
Indeed, it does make sure that an unallocated `raw` component of the `string_type` behaves like an empty string:
https://github.com/fortran-lang/stdlib/blob/bc37bcc8f71605f78e51c3c79c2d914e7825a339/src/stdlib_string_type.f90#L1166-L1177

**Comment by Aman-Godara** [comment]
May I ask why we didn't design the constructor to have an empty character sequence ("") when optional argument `string` is not given?
https://github.com/fortran-lang/stdlib/blob/bc37bcc8f71605f78e51c3c79c2d914e7825a339/src/stdlib_string_type.f90#L346-L352

**Comment by awvwgk** [comment]
There is still the possibility to have an uninitialized `string_type` which will have an unallocated `raw` component by default as in
```f90
use stdlib_string_type
implicit none
type(string_type) :: str
print *, str
end

**Comment by Aman-Godara** [comment]
Got it! 👍 Thanks!

**Pull request merged**